README.md
<h1>Terminus TSLint Frontend Configuration</h1>

[![CircleCI][circle-badge]][circle-link]
[![semantic-release][semantic-release-badge]][semantic-release]
[![MIT License][license-image]][license-url]
[![NPM version][npm-version-image]][npm-url]
[![Github release][gh-release-badge]][gh-releases]

A collection of TypeScript & Angular lint rules for Terminus frontend codebases.

> For ESLint configuration, see: https://github.com/GetTerminus/eslint-config-frontend


<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Installation](#installation)
- [Rulesets overview](#rulesets-overview)
- [Set up](#set-up)
  - [CI](#ci)
  - [Development](#development)
  - [Testing](#testing)
- [Rule Decisions](#rule-decisions)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->


## Installation

```
$ yarn add tslint @terminus/tslint-config-frontend -D
```

## Rulesets overview

There are 3 rulesets defined for TSLint:

1. **[CI](#CI)**: This is the default ruleset. This is meant to be used during your CI builds so it throws hard errors when issues are
   found.
2. **[Development](#Development)**: This enforces all the same rules as the CI ruleset but infractions are reported as warnings rather than
   errors.
3. **[Testing](#Testing)**: This ruleset extends the development ruleset and then further disables certain tests that make writing spec
   files less arduous (`no-non-null-assertion`, component naming requirements, etc).


## Set up

You will need to set up separate configs and scripts for each ruleset: `ci`, `development` and `testing`. Creating a custom script call for
each within your `package.json` will allow for easy composability of commands once all linters are set up.

### CI

[`CI Rules`](./ci.js)

#### 1. Create the file and extend our ruleset

Create a TSLint config file at the root level named `tslint.ci.json` and extend the base ruleset:

```json
{
  "extends": "@terminus/tslint-config-frontend"
}
```

> Linting during the CI process is the most strict and will fail if _any_ issues are found. The only way a linting issue makes it to CI is
> because someone didn't lint before commiting.

#### 2. Add a linting command to `package.json`

- The `--project` flag reference should point to the primary app `tsconfig` file.
- The `--config` flag reference should point to the ci `tslint` file.

```json
{
  "name": "My Project",
  "scripts": {
    "lint:tslint:ci": "npx tslint --project ./src/tsconfig.app.json --config ./tslint.ci.json"
  }
}
```


### Development

This ruleset extends the `ci` ruleset but softens the rules to turn many stylistic issues into warnings in order to not impede development.

[`Development Rules`](./development.js)

#### 1. Create the file and extend our ruleset

Create a TSLint config file at the root level named `tslint.json` and extend the development ruleset:

```json
{
  "extends": "@terminus/tslint-config-frontend/development"
}
```

#### 2. Add project specific rules

> NOTE: When adjusting a TSLint rule, the _entire_ rule must be defined again.

```json
{
  "extends": "@terminus/tslint-config-frontend/development",
  "rules": {
    "component-selector": [
      true,
      "element",
      "foo",
      "kebab-case"
    ],
    "directive-selector": [
      true,
      "attribute",
      "foo",
      "camelCase"
    ],
    "pipe-prefix": [
      true,
      "foo"
    ]
  }
}
```

#### 3. Add a linting command to `package.json`

After the `--project` flag, the reference should point to your primary app `tsconfig` file.

```json
{
  "name": "My Project",
  "scripts": {
    "lint:tslint": "npx tslint --project ./src/tsconfig.app.json"
  }
}
```


### Testing

[`Testing Rules`](./testing.js)

#### 1. Create the file and extend our ruleset

Create a TSLint config file at the root level named `tslint.spec.json` and extend the testing ruleset:

```json
{
  "extends": "@terminus/tslint-config-frontend/testing"
}
```

#### 2. Add a linting command to `package.json`

- The `--project` flag reference should point to the spec `tsconfig` file.
- The `--config` flag reference should point to the spec `tslint` file.

```json
{
  "name": "My Project",
  "scripts": {
    "lint:tslint:spec": "npx tslint --project ./src/tsconfig.spec.json --config ./tslint.spec.json"
  }
}
```


## Rule Decisions

Each rule is accompanied by a comment outlining the reasoning behind the decision to include the rule.

For most rules, see [`ci.js`](./ci.js).




<!--
  LINKS
-->

<!-- BADGES -->
[circle-badge]:           https://circleci.com/gh/GetTerminus/tslint-config-frontend/tree/release.svg?style=shield
[npm-version-image]:      http://img.shields.io/npm/v/@terminus/tslint-config-frontend.svg
[semantic-release-badge]: https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg
[license-image]:          http://img.shields.io/badge/license-MIT-blue.svg
[gh-release-badge]:       https://img.shields.io/github/release/GetTerminus/tslint-config-frontend.svg
[gh-releases]:            https://github.com/GetTerminus/tslint-config-frontend/releases/
[circle-link]:            https://circleci.com/gh/GetTerminus/tslint-config-frontend/tree/release
[semantic-release]:       https://github.com/semantic-release/semantic-release
[npm-url]:                https://npmjs.org/package/@terminus/tslint-config-frontend
[license-url]:            https://github.com/GetTerminus/tslint-config-frontend/blob/release/LICENSE


CHANGELOG.md
## <small>1.0.7 (2019-08-02)</small>

* fix: 🐛 no-negated-async template rule removed ([e2d13e8](https://github.com/GetTerminus/tslint-config-frontend/commit/e2d13e8)), closes [#26](https://github.com/GetTerminus/tslint-config-frontend/issues/26)
* chore(deps): bump lodash from 4.17.11 to 4.17.14 ([c035b1a](https://github.com/GetTerminus/tslint-config-frontend/commit/c035b1a))

## <small>1.0.6 (2019-06-11)</small>

* fix: 🐛 update peer dependencies for Angular 8 ([be064cd](https://github.com/GetTerminus/tslint-config-frontend/commit/be064cd))

## <small>1.0.5 (2019-06-11)</small>

* fix: 🐛 update dependencies ([ab935e9](https://github.com/GetTerminus/tslint-config-frontend/commit/ab935e9))
* chore(deps): bump fstream from 1.0.11 to 1.0.12 ([4e232fc](https://github.com/GetTerminus/tslint-config-frontend/commit/4e232fc))
* chore(deps): bump handlebars from 4.1.1 to 4.1.2 (#10) ([d64b35a](https://github.com/GetTerminus/tslint-config-frontend/commit/d64b35a)), closes [#10](https://github.com/GetTerminus/tslint-config-frontend/issues/10)

## <small>1.0.4 (2019-06-05)</small>

* fix: 🐛 spec rules now disables no-non-null-assertion ([9caa64c](https://github.com/GetTerminus/tslint-config-frontend/commit/9caa64c))
* Add codeowners file ([9326946](https://github.com/GetTerminus/tslint-config-frontend/commit/9326946))

## <small>1.0.3 (2019-04-30)</small>

* fix: rule updates after integrating into UI library ([82300b6](https://github.com/GetTerminus/tslint-config-frontend/commit/82300b6))

## <small>1.0.2 (2019-04-17)</small>

* fix: move peer deps to primary deps ([3712859](https://github.com/GetTerminus/tslint-config-frontend/commit/3712859))
* Replace master with release branches for badges ([2352a2b](https://github.com/GetTerminus/tslint-config-frontend/commit/2352a2b))
* Merge pull request #2 from GetTerminus/add-badges ([023955b](https://github.com/GetTerminus/tslint-config-frontend/commit/023955b)), closes [#2](https://github.com/GetTerminus/tslint-config-frontend/issues/2)
* chore: add badges to readme ([133543e](https://github.com/GetTerminus/tslint-config-frontend/commit/133543e))

## <small>1.0.1 (2019-04-16)</small>

* Merge pull request #1 from GetTerminus/update-version ([43aa126](https://github.com/GetTerminus/tslint-config-frontend/commit/43aa126)), closes [#1](https://github.com/GetTerminus/tslint-config-frontend/issues/1)
* fix: correctly set placeholder version ([f889e47](https://github.com/GetTerminus/tslint-config-frontend/commit/f889e47))

## 1.0.0 (2019-04-16)

* feat: First pass at all rules ([551d203](https://github.com/GetTerminus/tslint-config-frontend/commit/551d203))
* feat: Rulesets and Docs finalized ([d0b0056](https://github.com/GetTerminus/tslint-config-frontend/commit/d0b0056))


testing.js
module.exports = {
  'extends': './development.js',
  'env': {
    'jest': true,
  },
  'rules': {
    // Don't enforce naming conventions in test items
    'component-class-suffix': false,
    'directive-class-suffix': false,
    'pipe-prefix': false,

    // Allow inline templates
    'component-max-inline-declarations': false,

    // We often need to coerce to `any` in tests to verify how bad inputs are handled
    'no-any': false,
    'no-unsafe-any': false,

    // Allow lifecycle methods to be manually triggered
    'no-lifecycle-call': false,

    // Allow non-null assertions in spec files
    'no-non-null-assertion': false,

    // This allows easy access to private properties in tests
    'no-string-literal': false,

    // We aren't as concerned with test components performance
    'prefer-on-push-component-change-detection': false,

    // Don't enforce member-access for test components
    'prefer-output-readonly': false,

    // Don't require member access to be defined in test components
    'member-access': false,

    // Don't enforce accessibility for test components
    'template-accessibility-alt-text': false,
    'template-accessibility-elements-content': false,
    'template-accessibility-label-for': false,
    'template-accessibility-valid-aria': false,

    // Don't force trackBy in test components
    'template-use-track-by-function': false,

    // Selectors are rarely needed for test host components
    'use-component-selector': false,
  },
};


tooling/cz-config.js
'use strict';

module.exports = {

  'types': [
    {
      'value': 'feat',
      'name': 'feat: A new feature',
    },
    {
      'value': 'fix',
      'name': 'fix:      A bug fix',
    },
    {
      'value': 'docs',
      'name': 'docs:     Documentation only changes',
    },
    {
      'value': 'style',
      'name': 'style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)',
    },
    {
      'value': 'refactor',
      'name': 'refactor: A code change that neither fixes a bug nor adds a feature',
    },
    {
      'value': 'chore',
      'name': 'chore: Changes to the build process or auxiliary tools and libraries such as documentation generation',
    },
    {
      'value': 'revert',
      'name': 'revert: Revert to a commit',
    },
    {
      'value': 'WIP',
      'name': 'WIP: Work in progress',
    },
  ],

  'scopes': [
    {
      'name': 'CI Rules',
    },
    {
      'name': 'Dev Rules',
    },
    {
      'name': 'Spec Rules',
    },
    {
      'name': 'Build',
    },
  ],

  'allowCustomScopes': true,
  'allowBreakingChanges': [
    'feat',
    'fix',
    'refactor',
    'revert',
    'chore',
  ],

};


ci.js
module.exports = {

  // Extend existing rulesets
  'rulesDirectory': [
    'codelyzer',
  ],

  'defaultSeverity': 'error',

  // NOTE: Rules are tagged with the source: `#tslint` or `#codelyzer` and if it is fixable: `#hasFixer`
  'rules': {
    // An interface or literal type with just a call signature can be written as a function type.
    // #tslint #hasFixer
    'callable-types': true,

    // JavaScript and general programming convention is to refer to classes in PascalCase.
    // It can be confusing to use camelCase or other conventions for class names.
    // #tslint
    'class-name': true,

    // Helps to maintain a consistent, readable style in the codebase.
    // #tslint #hasFixer
    'comment-format': [
      true,
      'check-space',
    ],

    // Consistent conventions make it easy to quickly identify and reference assets of different types.
    // #codelyzer
    'component-class-suffix': true,

    // Large, inline templates and styles obscure the component’s purpose and implementation, reducing readability and maintainability.
    // #codelyzer
    'component-max-inline-declarations': [
      true,
      {
        'animations': 15,
        'styles': 3,
        'template': 3,
      },
    ],

    // a) Keeps the element names consistent with the specification for Custom Elements.
    // b) Consistent conventions make it easy to quickly identify and reference assets of different types.
    // c) Components are easy to identify in the DOM.
    // See https://angular.io/styleguide#style-02-07 https://angular.io/styleguide#style-05-02 https://angular.io/styleguide#style-05-02
    // #codelyzer
    'component-selector': [
      true,
      'element',
      // No prefix is enforced at the base level. This should be set specifically by consumers.
      '',
      'kebab-case',
    ],

    // Some decorators can only be used in certain class types. For example, an @Input should not be used in an @Injectable class.
    // #codelyzer
    'contextual-decorator': true,

    // Certain lifecycle methods can only be used in certain class types.
    // #codelyzer
    'contextual-lifecycle': true,

    // Prevent possible accidental execution due to missing brackets
    // #tslint #hasFixer
    'curly': 'error',

    // We should be alerted whenever using an identifier marked with a JSDoc `@deprecated` flag.
    // #tslint
    'deprecation': true,

    // Consistent conventions make it easy to quickly identify and reference assets of different types.
    // #codelyzer
    'directive-class-suffix': true,

    // a) Consistent conventions make it easy to quickly identify and reference assets of different types.
    // b) It is easier to recognize that a symbol is a directive by looking at the template’s HTML.
    // See https://angular.io/styleguide#style-02-06 https://angular.io/styleguide#style-02-08
    // #codelyzer
    'directive-selector': [
      true,
      'attribute',
      '',
      'camelCase',
    ],

    // For some libraries, importing the library directly can cause unused submodules to be loaded, so you may want to block these imports
    // and require that users directly import only the submodules they need. In other cases, you may simply want to ban an import because
    // using it is undesirable or unsafe.
    // #tslint
    'import-blacklist': [
      true,
      'lodash',
    ],

    // Imports are easier for the reader to look at when they’re tidy.
    // #codelyzer
    'import-destructuring-spacing': true,

    // Helps to maintain a consistent, readable style in the codebase.
    // #tslint
    'import-spacing': true,

    // Interfaces are generally preferred over type literals because interfaces can be implemented, extended and merged.
    // #tslint #hasFixer
    'interface-over-type-literal': true,

    // Helps to maintain a consistent, readable style in the codebase.
    // #tslint
    'jsdoc-format': [
      true,
      'check-multiline-start',
    ],

    // Labels in JavaScript only can be used in conjunction with break or continue, constructs meant to be used for loop flow control. While
    // you can theoretically use labels on any block statement in JS, it is considered poor code structure to do so.
    // #tslint
    'label-position': true,

    // Explicit visibility declarations can make code more readable and accessible for those new to TS. Members lacking a visibility
    // declaration may be an indication of an accidental leak of class internals.
    // #tslint #hasFixer
    'member-access': [
      true,
      'check-accessor',
      'check-parameter-property',
    ],

    // Both formats of type assertions have the same effect, but only as type assertions work in .tsx files.
    // #tslint #hasFixer
    'no-angle-bracket-type-assertion': true,

    // Using any as a type declaration nullifies the compile-time benefits of the type system.
    // #tslint
    'no-any': true,

    // @Attribute is considered bad practice. Use @Input instead.
    // #codelyzer
    'no-attribute-decorator': true,

    // Comparing boolean values to boolean literals is unnecessary, as those expressions will result in booleans too.
    // #tslint #hasFixer
    'no-boolean-literal-compare': true,

    // Ensure that directives do not implement conflicting lifecycle interfaces.
    // #codelyzer
    'no-conflicting-lifecycle': true,

    // The second call to ‘super()’ will fail at runtime.
    // #tslint
    'no-duplicate-super': true,

    // An empty interface is equivalent to its supertype (or `{}`).
    // #tslint
    'no-empty-interface': true,

    // The property associated with @HostBinding or the method associated with @HostListener can be modified only in a single place: in the
    // directive’s class. If you use the host metadata property, you must modify both the property declaration inside the controller, and
    // the metadata associated with the directive.
    // See: https://angular.io/styleguide#style-06-03
    // #codelyzer
    'no-host-metadata-property': true,

    // Disallows importing modules that are not listed as dependency in the project’s package.json
    // #tslint
    'no-implicit-dependencies': true,

    // Explicit types where they can be easily inferred by the compiler make code more verbose.
    // #tslint #hasFixer
    'no-inferrable-types': true,

    // a) It is easier and more readable to identify which properties in a class are inputs.
    // b) If you ever need to rename the property name associated with @Input, you can modify it in a single place.
    // c) The metadata declaration attached to the directive is shorter and thus more readable.
    // See https://angular.io/styleguide#style-05-12
    // #codelyzer
    'no-inputs-metadata-property': true,

    // Two names for the same property (one private, one public) is inherently confusing.
    // #codelyzer
    'no-input-rename': true,

    // Explicit calls to lifecycle methods could be confusing. Invoke them is an Angular’s responsability.
    // #codelyzer
    'no-lifecycle-call': true,

    // Interfaces in TypeScript aren’t meant to describe constructors on their implementations.
    // #tslint
    'no-misused-new': true,

    // Using non-null assertion cancels the benefits of the strict null checking mode.
    // #tslint
    'no-non-null-assertion': true,

    // Listeners subscribed to an output with such a name will also be invoked when the native event is raised.
    // #codelyzer
    'no-output-native': true,

    // Two names for the same property (one private, one public) is inherently confusing.
    // See https://angular.io/styleguide#style-05-13
    // #codelyzer
    'no-output-rename': true,

    // Angular allows for an alternative syntax on-*. If the event itself was prefixed with on this would result in an on-onEvent binding
    // expression.
    // See https://angular.io/guide/styleguide#dont-prefix-output-properties
    // #codelyzer
    'no-output-on-prefix': true,

    // a) It is easier and more readable to identify which properties in a class are events.
    // b) If you ever need to rename the event name associated with @Output, you can modify it in a single place.
    // c) The metadata declaration attached to the directive is shorter and thus more readable.
    // See https://angular.io/styleguide#style-05-12
    // #codelyzer
    'no-outputs-metadata-property': true,

    // Impure pipes do not perform well because they run on every change detection cycle.
    // #codelyzer
    'no-pipe-impure': true,

    // If you ever need to rename the property associated with @ContentChild, @ContentChildren, @ViewChild or @ViewChildren, you can modify
    // it in a single place.
    // #codelyzer
    'no-queries-metadata-property': true,

    // Warns if a type assertion does not change the type of an expression.
    // #tslint #hasFixer
    'no-unnecessary-type-assertion': true,

    // If you’re dealing with data of unknown or “any” types, you shouldn’t be accessing members of it. Either add type annotations for
    // properties that may exist or change the data type to the empty object type `{}`.
    // #tslint
    'no-unsafe-any': true,

    // Unused CSS is are likely an error due to incomplete refactoring.
    // #codelyzer #hasFixer
    'no-unused-css': true,

    // Helps to maintain a consistent, readable style in the codebase.
    // NOTE: ESLint has a related rule, but the 'fixer' does not fix sort order.
    // #tslint #hasFixer
    'ordered-imports': [
      true,
      {
        'case-insensitive': true,
      },
    ],

    // Consistent conventions make it easy to quickly identify and reference assets of different types.
    // #codelyzer
    // No prefix is enforced at the base level. This should be set specifically by consumers.
    'pipe-prefix': [
      true,
      '',
    ],

    // Enforces component’s change detection to ChangeDetectionStrategy.OnPush for better performance.
    // #codelyzer
    'prefer-on-push-component-change-detection': true,

    // Prefer to declare @Output as readonly since they should not be reassigned.
    // #codelyzer
    'prefer-output-readonly': true,

    // The ./ prefix is standard syntax for relative URLs; don’t depend on Angular’s current ability to do without that prefix.
    // See https://angular.io/guide/styleguide#style-05-04
    // #codelyzer
    'relative-url-prefix': true,

    // Adding two variables not of the same type can easily cause unexpected behavior.
    // #tslint
    'restrict-plus-operands': true,

    // Prefer return; in void functions and return undefined; in value-returning functions.
    // #tslint
    'return-undefined': true,

    // Helps to maintain a consistent, readable style in the codebase.
    // NOTE: The eslint fixer doesn't seem to work for this one so we are using TSLint's semicolon enforcement.
    // #tslint #hasFixer
    'semicolon': [
      true,
      'always',
    ],

    // Enforces alternate text for elements which require the alt, aria-label, aria-labelledby attributes for better accessibility
    // #codelyzer
    'template-accessibility-alt-text': true,

    // Heading, anchor and button elements should have content to be accessible by screen readers
    // #codelyzer
    'template-accessibility-elements-content': true,

    // The label tag should either have a `for` attribute or should have associated control.
    // #codelyzer
    'template-accessibility-label-for': true,

    // Ensures that the correct ARIA attributes are used.
    // #codelyzer
    'template-accessibility-valid-aria': true,

    // Ensure that the two-way data binding syntax is correct.
    // #codelyzer #hasFixer
    'template-banana-in-box': true,

    // The use of ‘$any’ nullifies the compile-time benefits of the Angular’s type system.
    // #codelyzer
    'template-no-any': true,

    // Calling expressions in templates causes it to run on every change detection cycle and may cause performance issues.
    // #codelyzer
    'template-no-call-expression': true,

    // Elements that can be visually distracting can cause accessibility issues with visually impaired users (e.g. `blink` or `marquee`).
    // #codelyzer
    'template-no-distracting-elements': true,

    // The use of ‘trackBy’ is considered a good practice and is better for performance.
    // #codelyzer
    'template-use-track-by-function': true,

    // Helps to maintain a consistent, readable style in the codebase.
    // #hasFixer
    'type-literal-delimiter': true,

    // Any two overloads that could be unified into one by using a union or an optional/rest parameter should be.
    // #tslint
    'unified-signatures': true,

    // Omitting the component selector makes debugging difficult.
    // #codelyzer
    'use-component-selector': true,

    // Interfaces prescribe typed method signatures. Use those signatures to flag spelling and syntax mistakes.
    // #codelyzer
    'use-lifecycle-interface': true,

    // Interfaces prescribe typed method signatures. Use those signatures to flag spelling and syntax mistakes.
    // #codelyzer
    'use-pipe-decorator': true,

    // Interfaces prescribe typed method signatures. Use those signatures to flag spelling and syntax mistakes.
    // #codelyzer
    'use-pipe-transform-interface': true,
  },
};
